### DISCUSSION TITLE
mcp-golang - A go SDK for implementing MCP servers in go quickly

### DISCUSSION THREAD

**Discussion by Chrisbattarbee** [modelcontextprotocol/.github#85] [Show and tell]
### Pre-submission Checklist

- [X] This post follows the MCP community guidelines

### What would you like to share?

I recently open sourced [mcp-golang](https://github.com/metoro-io/mcp-golang). A high-level unofficial sdk for building MCP servers in golang.

The main focus is maintaining type safety and auto generating as much as we can from type definitions. 

We use reflection to figure out the args of the tool/resource/prompt handlers and use this to autogenerate the jsonschema which we serve to clients when they list resources. Auto serialize / deserialize, handle errors etc.

Quick example of this below in an example weather server.

```go
package main

import (
	"fmt"
	mcp_golang "github.com/metoro-io/mcp-golang"
	"github.com/metoro-io/mcp-golang/transport/stdio"
	"io"
	"net/http"
)

type WeatherArguments struct {
	Longitude float64 `json:"longitude" jsonschema:"required,description=The longitude of the location to get the weather for"`
	Latitude  float64 `json:"latitude" jsonschema:"required,description=The latitude of the location to get the weather for"`
}

// This is explained in the docs at https://mcpgolang.com/tools
func main() {
	done := make(chan struct{})
	server := mcp_golang.NewServer(stdio.NewStdioServerTransport())
	err := server.RegisterTool("get_weather", "Get the weather forecast for temperature, wind speed and relative humidity", func(arguments WeatherArguments) (*mcp_golang.ToolResponse, error) {
		url := fmt.Sprintf("https://api.open-meteo.com/v1/forecast?latitude=%f&longitude=%f&current=temperature_2m,wind_speed_10m&hourly=temperature_2m,relative_humidity_2m,wind_speed_10m", arguments.Latitude, arguments.Longitude)
		resp, err := http.Get(url)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		output, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}
		return mcp_golang.NewToolReponse(mcp_golang.NewTextContent(string(output))), nil
	})
	err = server.Serve()
	if err != nil {
		panic(err)
	}
	<-done
}

```
From the type definitions we can generate out the following `tools/list` endpoint.

```
client:  {"method":"tools/list","params":{},"jsonrpc":"2.0","id":1}
server: {"id":1,"jsonrpc":"2.0","result":{"tools":[{"description":"Get the weather forecast for temperature, wind speed and relative humidity","inputSchema":{"$schema":"https://json-schema.org/draft/2020-12/schema","properties":{"longitude":{"type":"number","description":"The longitude of the location to get the weather for"},"latitude":{"type":"number","description":"The latitude of the location to get the weather for"}},"type":"object","required":["longitude","latitude"]},"name":"get_weather"}]}}
``` 

Demo of it in action with Claude!

https://github.com/user-attachments/assets/cd78ce39-872d-4401-a80f-a826afdf29b6

We've currently implemented the required functionality for servers. We're working through the optional features now, with the idea of supporting clients after that. Any help is appreciated!

### Relevant Links

https://github.com/metoro-io/mcp-golang

**Comment by justinclift-prvidr** [comment]
Just to check, have you seen this discussion of an official MCP Go SDK from the Go team?

https://github.com/orgs/modelcontextprotocol/discussions/364
